be/src/storage/lake/transactions.cpp

```diff
@@ -51,42 +51,60 @@ StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t t
     }
 
     auto base_metadata = std::move(base_metadata_or).value();
-    auto new_metadata = std::make_shared<TabletMetadataPB>(*base_metadata);
-    auto log_applier = new_txn_log_applier(Tablet(tablet_mgr, tablet_id), new_metadata, new_version);
-
-    if (new_metadata->compaction_inputs_size() > 0) {
-        new_metadata->mutable_compaction_inputs()->Clear();
-    }
-
-    if (new_metadata->orphan_files_size() > 0) {
-        new_metadata->mutable_orphan_files()->Clear();
-    }
-
-    if (base_metadata->compaction_inputs_size() > 0 || base_metadata->orphan_files_size() > 0) {
-        new_metadata->set_prev_garbage_version(base_metadata->version());
-    }
-
-    new_metadata->set_commit_time(commit_time);
-
-    auto init_st = log_applier->init();
-    if (!init_st.ok()) {
-        if (init_st.is_already_exist()) {
-            return new_version_metadata_or_error(init_st);
-        } else {
-            return init_st;
-        }
-    }
-
+    std::unique_ptr<TxnLogApplier> log_applier;
+    std::shared_ptr<TabletMetadataPB> new_metadata;
     std::vector<std::string> files_to_delete;
 
     // Apply txn logs
     int64_t alter_version = -1;
-    for (auto txn_id : txn_ids) {
+    // Do not delete txn logs if txns_size != 1, let gc do the work
+    // If the txn logs are deleted, it will be tricky to handle the situation of batch publish switching to single.
+
+    // for example:
+    // 1. the mode of publish is batch,
+    // 2. txn2 and txn3 have been published successfully and visible version in FE is updated to 3,
+    // 3. then txn4 and txn5 are published successfully in BE and the txn_log of txn4 and txn5 have been deleted,
+    // but FE do not get the response for some reason,
+    // 4. turn the mode of publish to single,
+    // 5. txn4 will be published in later publish task, but we can't judge what's the latest_version in BE and we can not reapply txn_log if
+    // txn logs have been deleted.
+    bool delete_txn_log = (txn_ids.size() == 1);
+    for (int i = 0; i < txn_ids.size(); i++) {
```

Comment:
This for loop body is so long, could you please break it into some smaller methods?

Reply:
I will refactor the code when optimizing  the gc of txn_log